golang/sig0/doh.go

```diff
@@ -39,3 +46,79 @@ func SendUDPQuery(server string, m *dns.Msg) (*dns.Msg, error) {
 
 	return answer, nil
 }
+
+func FindDOHEndpoint(name string) (*url.URL, error) {
+	lookup := dns.Fqdn("_dns." + name)
+
+	svcbQry, err := QueryWithType(lookup, dns.TypeSVCB)
+	if err != nil {
+		return nil, fmt.Errorf("findDOHEndpoint: failed to construct query for %s: %w", lookup, err)
+	}
+
+	answer, err := SendDOHQuery(DefaultDOHResolver, svcbQry)
+	if err != nil {
+		return nil, fmt.Errorf("findDOHEndpoint: no answer for svcb query for %s: %w", lookup, err)
+	}
+	if len(answer.Answer) < 1 {
+		spew.Dump(answer)
+		return nil, fmt.Errorf("findDOHEndpoint: no answer section for %s", lookup)
+	}
+
+	first := answer.Answer[0]
+	svcb, ok := first.(*dns.SVCB)
+	if !ok {
+		return nil, fmt.Errorf("findDOHEndpoint: unable to cast %T answer for %s to SVCB", first, lookup)
+	}
+
+	if !dns.IsFqdn(svcb.Target) {
+		return nil, fmt.Errorf("findDOHEndpoint: Expected SVCB target to be FQDN but got %q", svcb.Target)
+	}
+	host := svcb.Target[:len(svcb.Target)-1]
+
+	var alpn *dns.SVCBAlpn
+	var dohPath *dns.SVCBDoHPath
+
+	for _, v := range svcb.Value {
+		switch tv := v.(type) {
+		case *dns.SVCBAlpn:
+			alpn = tv
+		case *dns.SVCBDoHPath:
+			dohPath = tv
+		default:
+			fmt.Fprintf(os.Stderr, "findDOHEndpoint: unexpected keyValue pair type: %T\n", tv)
+		}
+	}
+
+	if alpn == nil {
+		return nil, fmt.Errorf("findDOHEndpoint: expected alpn in svcb.Value")
+	}
+	if dohPath == nil {
+		return nil, fmt.Errorf("findDOHEndpoint: expected dohPath in svcb.Value")
+	}
+	if dohPath.Template == "" {
+		return nil, fmt.Errorf("findDOHEndpoint: expected dohPath Template value")
+	}
+
+	// TODO: write this more genericly but striclty speaky we should check for IN HTTPS to check if we can use https
+	var hasHttp = false
+	for _, a := range alpn.Alpn {
+		if strings.Contains(a, "h1") || strings.Contains(a, "h2") {
+			hasHttp = true
+		}
```

Comment:
doesn't check for http/3 yet but only supporting that seems a little fringe to me.

Reply:
but then again, i'm not even sure Go supports http/3 yet.